cpp/models/ode_secirts/parameters_io.cpp

```diff
@@ -0,0 +1,344 @@
+                    // but the algorithm is not robust enough to avoid large negative
+                    // values and there are tests that rely on it
+                    log_error("{:s} for age group {:s} is {:.4f} for region {:d}, "
+                              "exceeds expected negative value.",
+                              str, ConfirmedCasesDataEntry::age_group_names[i], value, region);
+                    value = 0.0;
+                }
+                else if (value < 0) {
+                    log_info("{:s} for age group {:s} is {:.4f} for region {:d}, "
+                             "automatically corrected",
+                             str, ConfirmedCasesDataEntry::age_group_names[i], value, region);
+                    value = 0.0;
+                }
+            };
+
+            const double tol_error = -1e-8;
+            try_fix_constraints(num_InfectedSymptoms[i], tol_error, "InfectedSymptoms");
+            try_fix_constraints(num_InfectedNoSymptoms[i], tol_error, "InfectedNoSymptoms");
+            try_fix_constraints(num_Exposed[i], tol_error, "Exposed");
+            try_fix_constraints(num_InfectedSevere[i], tol_error, "InfectedSevere");
+            try_fix_constraints(num_death[i], tol_error, "Dead");
+            try_fix_constraints(num_icu[i], tol_error, "InfectedCritical");
+            try_fix_constraints(num_timm_i[i], tol_error, "Recently Recovered");
+        }
+    }
+
+    return success();
+}
+
+IOResult<std::vector<std::vector<double>>> read_immunity_population(const std::string& path,
+                                                                    const size_t& num_age_groups)
+{
+    std::vector<std::vector<double>> ans(3, std::vector<double>(num_age_groups, 0.0));
+    std::fstream immunity_file;
+    immunity_file.open(path, std::ios::in);
+    if (immunity_file.fail()) { // checks to see if file opended
+        return mio::failure(mio::StatusCode::InvalidValue, "Failed to open immunity_population.txt.");
+    }
+    if (immunity_file.is_open()) {
+        std::string tp;
+        int linenumber = 0;
+        while (linenumber < 3) {
+            getline(immunity_file, tp);
+            // Find the last character that is not a line ending
+            size_t lastChar = tp.find_last_not_of("\r\n");
+            // If such a character was found, reduce the string
+            if (lastChar != std::string::npos) {
+                tp = tp.substr(0, lastChar + 1);
+            }
+            auto line = split(tp, ' ');
+            for (size_t i = 0; i < num_age_groups; i++) {
+                ans[linenumber][i] = std::stod(line[i]);
+            }
+            linenumber++;
+        }
+        immunity_file.close(); //close the file object.
+    }
+
+    // assert that the sum of the cols is 1 (tolerance 1-10)
+    for (size_t i = 0; i < 6; ++i) {
+        assert(std::abs(ans[0][i] + ans[1][i] + ans[2][i] - 1.0) < 1e-10);
+    }
+
+    return ans;
+}
```

Comment:
I think the entire immunity_(of_?)population should be replaced by a parameter. If you need to read or write it you can still use the serialization of MatrixXd, and you can use the get_default to store your values.
But it has to change from this. One configuration file with mystery numbers is tolerable, but the same file in three different locations is not acceptable.